### ISSUE TITLE
Self referencing loop detected

### ISSUE THREAD

**Issue by Discjoggy** [Xabaril/AspNetCore.Diagnostics.HealthChecks#16]
Hi,

the response of GET requests produces an error response when a services is unavailable which is then unusable, e.g. for a health check ui service. First I used the healt check support for `AzureStorageEmulator` and I thought this is an emulator specific issue, but I've also tried the provided health check support for `Identity Server` and `Uri` and everything is working well as long as the services are available - when it is not, a self referencing loop was detected.

```json
{
  "error": {
    "code": "500",
    "message": "Self referencing loop detected for property 'exception' with type 'Microsoft.WindowsAzure.Storage.StorageException'. Path  'entries.azureblob.exception.requestInformation'."
  }
}
```

The `ConfigureServices` method contains this:
```cs
services.AddHealthChecks().AddAzureBlobStorage(configuration.GetConnectionString("AzureStorage"));
````
`
The `Configure` method looks like this:
```cs
app.UseHealthChecks("/api/health", new HealthCheckOptions()
{
    Predicate = _ => true,
    ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
});
```

**Comment by unaizorrilla** [comment]
i try to review this but probably the responde writer dont have the IgnoreLoop property, let me a few hours to check and update the package

**Comment by CarlosLanderas** [comment]
Hello @Discjoggy, when we ported BeatPulse to Asp.Net Health Checks we did not set properly the serialization settings. I was able to reproduce your problem and pushed the [commit ](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/commit/8d080a9f8960d9ce5e8b1e4dcd56dc7067662163).

UI.Client package version 2.2.1 is gonna get published in Nuget so if you can tell us when is indexed if is working for you we can close the issue.

Thank you very much opening this issue as this package is transversal and all the third party client libraries were affected by self referenced members

**Comment by unaizorrilla** [comment]
Thanks @CarlosLanderas 

@Discjoggy please clase the issue of the new packge solve the problem

**Comment by CarlosLanderas** [comment]
@Discjoggy  version 2.2.1 of the UI.Client is up

[https://www.nuget.org/packages/AspNetCore.HealthChecks.UI.Client/2.2.1](https://www.nuget.org/packages/AspNetCore.HealthChecks.UI.Client/2.2.1)

**Comment by Discjoggy** [comment]
Stil not working, sorry. It stil seems that Azure, Url and IdentityProvider have the same issue. 
But now the responses are different, I've tested 3 scenarios:

1. Every service is available: **Everything works as expected**

2. A service was never available: **A very long (10 MB !) response is returned**
```
{"entries":{"identityProvider4":{"data":{},"duration":"00:00:02.2186441","exception":{"message":"Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte","data":{},"innerException":{"ClassName":"System.Net.Sockets.SocketException","Message":"Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2147467259,"Source":"System.Private.CoreLib","WatsonBuckets":null,"NativeErrorCode":10061},"targetSite":{"name":"MoveNext","declaringType":"System.Net.Http.ConnectHelper+<ConnectAsync>d__2, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","reflectedType":"System.Net.Http.ConnectHelper+<ConnectAsync>d__2, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","memberType":"Method","metadataToken":100664484,"module":{"mdStreamVersion":131072,"fullyQualifiedName":"C:\\Program Files\\dotnet\\shared\\Microsoft.NETCore.App\\2.2.0\\System.Net.Http.dll","moduleVersionId":"ae655ede-cff8-4af3-86a3-dba3713c2805","metadataToken":1,"scopeName":"System.Net.Http.dll","name":"System.Net.Http.dll","assembly":{"codeBase":"file:///C:/Program Files/dotnet/shared/Microsoft.NETCore.App/2.2.0/System.Net.Http.dll","fullName":"System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","definedTypes":["<>f__AnonymousType0`1, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","<>f__AnonymousType1`3, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","<>f__AnonymousType2`2, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","<>f__AnonymousType3`3, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","<>f__AnonymousType4`4, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SECURITY_STATUS, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Kernel32, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+CredHandle, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+ContextFlags, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+Endianness, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+CredentialUse, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+SecBuffer, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+SspiCli+SecBufferDesc, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Crypt32, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Crypt32+CERT_CONTEXT, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Crypt32+SSL_EXTRA_CERT_CHAIN_POLICY_PARA, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Crypt32+CERT_CHAIN_POLICY_PARA, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+Crypt32+CERT_CHAIN_POLICY_STATUS, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a","Interop+WinHttp, System.Net.Http, Version=4.2.1.0, Culture=neutral, [...]
```

3. A service was available but then gets unavaiable: **The service hangs 2-3 minutes and then produces also a long response**

```
{
  "entries":{
    "identityProvider4":{
      "data":{

      },
      "duration":"00:01:40.0583760",
      "exception":{
        "ClassName":"System.Threading.Tasks.TaskCanceledException",
        "Message":"The operation was canceled.",
        "Data":null,
        "InnerException":{
          "ClassName":"System.IO.IOException",
          "Message":"Unable to read data from the transport connection: Der E/A-Vorgang wurde wegen eines Threadendes oder einer Anwendungsanforderung abgebrochen.",
          "Data":null,
          "InnerException":{
            "ClassName":"System.Net.Sockets.SocketException",
            "Message":"Der E/A-Vorgang wurde wegen eines Threadendes oder einer Anwendungsanforderung abgebrochen",
            "Data":null,
            "InnerException":null,
            "HelpURL":null,
            "StackTraceString":null,
            "RemoteStackTraceString":null,
            "RemoteStackIndex":0,
            "ExceptionMethod":null,
            "HResult":-2147467259,
            "Source":null,
            "WatsonBuckets":null,
            "NativeErrorCode":995
          },
          "HelpURL":null,
          "StackTraceString":"   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error)\r\n   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)\r\n   at System.Net.Http.HttpConnection.FillAsync()\r\n   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)\r\n   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)",
          "RemoteStackTraceString":null,
          "RemoteStackIndex":0,
          "ExceptionMethod":null,
          "HResult":-2146232800,
          "Source":"System.Net.Sockets",
          "WatsonBuckets":null
        },
        "HelpURL":null,
        "StackTraceString":"   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)\r\n   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\r\n   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\r\n   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\r\n   at HealthChecks.IdSvr.IdSvrHealthCheck.CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken)",
        "RemoteStackTraceString":null,
        "RemoteStackIndex":0,
        "ExceptionMethod":null,
        "HResult":-2146233029,
        "Source":"System.Net.Http",
        "WatsonBuckets":null
      },
      "status":"Unhealthy"
    },
    "workingSet Memory":{
      "data":{

      },
      "duration":"00:00:00.0034504",
      "status":"Healthy"
    },
    "private Memory":{
      "data":{

      },
      "duration":"00:00:00.0043947",
      "status":"Healthy"
    },
    "virtual Memory":{
      "data":{

      },
      "duration":"00:00:00.0035859",
      "status":"Healthy"
    },
    "mem":{
      "data":{

      },
      "duration":"00:00:00.0080460",
      "status":"Healthy"
    }
  },
  "status":"Unhealthy",
  "totalDuration":"00:01:40.0938550"
}
```

**Comment by unaizorrilla** [comment]
Yeap this is the issue, HealthChecksResult contains the whole exception.

Let us check a workaround for this

**Comment by unaizorrilla** [comment]
@Discjoggy,

A new package **2.2.2** for UI.Client and UI are right now on Nuget, can you test this version and check if the issue is solved?

Thanks again for fill this issue

**Comment by Discjoggy** [comment]
Well done ! Now it seems everything is working as excpected. Here a response for an unavailable IdentityProvider:

```
{
  "status":"Unhealthy",
  "totalDuration":"00:00:02.1390104",
  "entries":{
    "identityProvider4":{
      "data":{

      },
      "description":"Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte",
      "duration":"00:00:02.1057827",
      "exception":"Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte",
      "status":"Unhealthy"
    },
    "workingSet Memory":{
      "data":{

      },
      "duration":"00:00:00.0037678",
      "status":"Healthy"
    },
    "private Memory":{
      "data":{

      },
      "duration":"00:00:00.0040960",
      "status":"Healthy"
    },
    "virtual Memory":{
      "data":{

      },
      "duration":"00:00:00.0062819",
      "status":"Healthy"
    },
    "mem":{
      "data":{

      },
      "duration":"00:00:00.0019460",
      "status":"Healthy"
    }
  }
}
```

Is it possible to set the maximum timeout / waiting time for a healthcheck ? I am asking because, as you can see, the IdentityProvider check takes 2 seconds, what happens to multiple dependent services which gets checked via HTTP requests. May an enhancement feature?

Anyway this issue looks solved for me, thanks! :)

**Comment by unaizorrilla** [comment]
For timeout fill issue on https://github.com/aspnet/Diagnostics

I close this issue, reopen if you need
